### PR TITLE
Feature/5127 langugage preference

### PR DESCRIPTION
**Pull request by brindasasi**
## 👋 Introduction
🤖 Resolves {#5127}

## 🕵️ Details
This pr adds 2 new fields "preferred_language_for_interview" ,  "preferred_language_for_exam" into db and basically everywhere preferred language has been added , these 2 new fields are also added. 

Existing `Preferred Language` has been renamed to `Preferred Communication Language`
New field labels `Preferred Spoken Interview Language`  and `Preferred Written Exam Language`

## 🧪 Testing
Verify all the below forms showing the new drop downs as per the screen shot , able to edit , save and retain the saved values. 

User

- [ ] About Me Section
- [ ] About Me Form

Admin 
- [ ] Users table
- [ ] Pool Candidates table
- [ ] CSV download of Pool Candidates 
- [ ] Create User
- [ ] Update User

## 📸 Screenshot
Screen Shot 2023-01-10 at 10 49 54 PM
Screen Shot 2023-01-10 at 10 49 41 PM

Admin/users

Screen Shot 2023-01-11 at 1 57 33 PM

### CODE REVIEW

File: frontend/talentsearch/src/js/components/createAccount/createAccountOperations.graphql
Comment (brindasasi) [inline frontend/talentsearch/src/js/components/createAccount/createAccountOperations.graphql:33]: - [ ] revert this as we don't need it

**Comment by brindasasi** [comment]
translations strings are pending. But rest would be good for review now

**Review by tristan-orourke — COMMENTED** [review]
The two new language fields are required, in the Personal Information form. Should we add them to the isProfileComplete check? And if we do that, should we add a line to the migration which sets the new field to match the existing preferred_lang field for existing users?

File: frontend/admin/src/js/components/user/UserTable.tsx
Comment (tristan-orourke) [inline frontend/admin/src/js/components/user/UserTable.tsx:346]: I'm not entirely sure how sortColumnName works, but is there any reason its in all-caps here and not on other columns?

**Review by tristan-orourke — COMMENTED** [review]
Haven't tested it yet, but everything looks good so far on a read-through of the code 👍

File: frontend/admin/src/js/components/user/UserTable.tsx
Comment (brindasasi) [reply frontend/admin/src/js/components/user/UserTable.tsx:346]: no .. thats by mistake ... good catch!

**Comment by brindasasi** [comment]
> The two new language fields are required, in the Personal Information form. Should we add them to the isProfileComplete check? And if we do that, should we add a line to the migration which sets the new field to match the existing preferred_lang field for existing users?

yes. for sure. My understanding is that it should be in the profile complete check .

For existing users , it wouldn't affect the existing applications. Only on the new application they would be demanded to fill in the new values. Wouldn't that be ok? It would also be another reminder to tell them they can choose these values.

If not liking that approach, we could force update these values with preferred_lang value with migration. I'm just wondering if that would be defeating the purpose of having them to choose. But you know lets ask @Jerryescandon  here 😅 .

**Comment by brindasasi** [comment]
> Haven't tested it yet, but everything looks good so far on a read-through of the code 👍

thanks! btw, where am I missing that profileCompleteCheck .. could you point me out?

**Comment by Jerryescandon** [comment]
Just add them to the check moving forward, dont worry about migrating or retrofitting, DCM can manually ask people if they need to. Let just focus on building things now for future applications.

**Comment by brindasasi** [comment]
pool candidates csv 
[pool_candidates_2023-01-12T19_50_32.519Z.csv](https://github.com/GCTC-NTGC/gc-digital-talent/files/10405491/pool_candidates_2023-01-12T19_50_32.519Z.csv)

**Comment by tristan-orourke** [comment]
I just want to note that if we add these new fields to the profileIsComplete check, and we do not set default values for existing users, then any existing users who already completed their profiles will no longer be considered complete.

**Comment by Jerryescandon** [comment]
> I just want to note that if we add these new fields to the profileIsComplete check, and we do not set default values for existing users, then any existing users who already completed their profiles will no longer be considered complete.

:NotLikeThis:

File: api/app/GraphQL/Mutations/PoolCandidateSnapshot.graphql
Comment (vd1992) [inline api/app/GraphQL/Mutations/PoolCandidateSnapshot.graphql:13]: I don't think this should be removed?

**Comment by vd1992** [comment]
> I just want to note that if we add these new fields to the profileIsComplete check, and we do not set default values for existing users, then any existing users who already completed their profiles will no longer be considered complete.

in the migration, after the new columns are created, I think it would be reasonable to just default assign the value of `preferred_lang` to them
users can change them to their liking afterwards if they prefer, but profile completeness won't be broken

something like this, or however the SQL needs to be formatted

```
        DB::statement(
            <<<SQL
                UPDATE users
                    SET preferred_language_for_interview =
                        case preferred_lang
                            when :english then :english
                            when :french then :french
                        end
            SQL, [
                'english' =>'en',
                'french' => 'fr,
            ]
        );
```

File: api/app/GraphQL/Mutations/PoolCandidateSnapshot.graphql
Comment (brindasasi) [reply api/app/GraphQL/Mutations/PoolCandidateSnapshot.graphql:13]: oops .. thank you!

**Comment by brindasasi** [comment]
> > I just want to note that if we add these new fields to the profileIsComplete check, and we do not set default values for existing users, then any existing users who already completed their profiles will no longer be considered complete.
> 
> :NotLikeThis:



> > I just want to note that if we add these new fields to the profileIsComplete check, and we do not set default values for existing users, then any existing users who already completed their profiles will no longer be considered complete.
> 
> in the migration, after the new columns are created, I think it would be reasonable to just default assign the value of `preferred_lang` to them users can change them to their liking afterwards if they prefer, but profile completeness won't be broken
> 
> something like this, or however the SQL needs to be formatted
> 
> ```
>         DB::statement(
>             <<<SQL
>                 UPDATE users
>                     SET preferred_language_for_interview =
>                         case preferred_lang
>                             when :english then :english
>                             when :french then :french
>                         end
>             SQL, [
>                 'english' =>'en',
>                 'french' => 'fr,
>             ]
>         );
> ```



> > I just want to note that if we add these new fields to the profileIsComplete check, and we do not set default values for existing users, then any existing users who already completed their profiles will no longer be considered complete.
> 
> :NotLikeThis:

@Jerryescandon how do you want to go about this? For existing users, should we assign communication language to other two as well ?

**Comment by Jerryescandon** [comment]
I agree, I feel like its a fair assumption at this point to inherit the communication language selection and not break things, later on we (and by we I mean DCM) can send emails to let people know about these fields and prompt them to update if they so wish to.

**Comment by brindasasi** [comment]
> > I just want to note that if we add these new fields to the profileIsComplete check, and we do not set default values for existing users, then any existing users who already completed their profiles will no longer be considered complete.
> 
> in the migration, after the new columns are created, I think it would be reasonable to just default assign the value of `preferred_lang` to them users can change them to their liking afterwards if they prefer, but profile completeness won't be broken
> 
> something like this, or however the SQL needs to be formatted
> 
> ```
>         DB::statement(
>             <<<SQL
>                 UPDATE users
>                     SET preferred_language_for_interview =
>                         case preferred_lang
>                             when :english then :english
>                             when :french then :french
>                         end
>             SQL, [
>                 'english' =>'en',
>                 'french' => 'fr,
>             ]
>         );
> ```

I just gave the other column name for that .. It seems to be working fine when tested

**Comment by brindasasi** [comment]
After the latest main merge I get this error. Not sure why it would remove the french translations. Looking into it 
Screen Shot 2023-01-18 at 10 56 10 AM

**Comment by vd1992** [comment]
warning observed on the users table

image

**Comment by vd1992** [comment]
table columns, both toggle despite only clicking on one of them

gif

**Comment by vd1992** [comment]
for download csv and profile print
on **USERS** table
language is missing, only one that appears is preference for english/french/bilingual column

image

**Comment by vd1992** [comment]
snapshot is not capturing the new fields or not retrieving from the snapshot, I submitted an application on this branch after migrating

image

**Review by vd1992 — CHANGES_REQUESTED** [review]
please see the above series of comments and pictures

most of the problems seem localized to the _users_ table, _pool candidate_ table appears to work 
csv download and table columns for example are only throwing issues on the _users_ table

**Comment by brindasasi** [comment]
> language preferences are not reflected in the about me for candidate profile tab at `/admin/users/:id` thus they are not shown in the print either, I think they ought to be
> 
> image
> 
> image

This form shouldn't be affected. There is no language preference shown already.

**Comment by brindasasi** [comment]
> for download csv and profile print on **USERS** table language is missing, only one that appears is preference for english/french/bilingual column
> 
> image

There shouldn't be csv download change on users table as well as I mentioned in the check list above. The language that is showing here is not the langugage preference. It is about the jobs/opportunities that the applicant willing to work on.

**Comment by brindasasi** [comment]
> 

 label id looks to be the same which is causing this issue. I think I messed up after trans or merging.

**Comment by brindasasi** [comment]
> warning observed on the users table
> 
> image

Seems to be the root cause for below issue on the table

**Comment by brindasasi** [comment]
> does the snapshot view work on your end? it doesn't for me and I don't get why since I can see it is passing in the new fields
> 
> image

is this a old snapshot?  🤔

**Comment by brindasasi** [comment]
> does the snapshot view work on your end? it doesn't for me and I don't get why since I can see it is passing in the new fields
> 
> image
It should be good now. Could you check? 
Screen Shot 2023-01-19 at 4 33 01 PM

**Comment by vd1992** [comment]
yep snapshots parse properly now!

**Review by vd1992 — APPROVED** [review]
not sure if @tristan-orourke is or will be reviewing this...

but, things look good to me now, my feedback has been addressed
great job!

**Comment by brindasasi** [comment]
> not sure if @tristan-orourke is or will be reviewing this...
> 
> but, things look good to me now, my feedback has been addressed great job!

Thanks Vachan! Will leave it until tmo. for merge

**Comment by tristan-orourke** [comment]
Thanks, I think @vd1992 has this covered, don't wait for my review! 😄

**Pull request merged**
